Decouple trial license resets from product version

AthenaEryma opened the pull request:
Prior to this commit, a new trial license was allowed for a cluster for each major version. This was a reasonable strategy when we were releasing major versions on a highly regular cadence, but since those releases have slowed, we may wish to offer new trial licenses that are not tied to a major version release.

This commit introduces the concept of a Trial Licence Era, which is independent from the Elasticsearch version. In order to avoid complications with serialization, the TrialLicenseEra class is serialized in a way that is wire-compatible with the general Version class, and maintains the existing behavior for trials started in released versions. Thorough unit tests are included to verify that TrialLicenseEra and Version are compatible in both streamed and XContent representations.

AthenaEryma committed:
Decouple trial license resets from product version

Prior to this commit, a new trial license was allowed for a
cluster for each major version. This was a reasonable strategy
when we were releasing major versions on a highly regular cadence,
but since those releases have slowed, we may wish to offer new
trial licenses that are not tied to a major version release.

This commit introduces the concept of a Trial Licence Era, which is
independent from the Elasticsearch version. In order to avoid
complications with serialization, the TrialLicenseEra class is serialized
in a way that is wire-compatible with the general Version class,
and maintains the existing behavior for trials started in released versions.
Thorough unit tests are included to verify that TrialLicenseEra and
Version are compatible in both streamed and XContent representations.

AthenaEryma committed:
Clean up some missed version -> era

AthenaEryma committed:
Correct tests & add clarifying comments

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into license/decouple-trial-from-version

elasticsearchmachine commented:
Pinging @elastic/es-security (Team:Security)

jakelandis commented on x-pack/plugin/core/src/test/java/org/elasticsearch/license/internal/TrialLicenseEraTests.java:
```diff
@@ -0,0 +1,98 @@
+    // This simulates a node of a version that supports transport eras receiving a message which is still using old-school Versions
+    private TrialLicenseEra versionToEraSerialization(Version version) throws IOException {
+        try (BytesStreamOutput output = new BytesStreamOutput()) {
+            output.setTransportVersion(TransportVersion.current());
```
Should the test also include Transport version prior to the introduction of `TransportVersion` ?

gwbrown replied:
I don't think that using a different version there changes anything in this test; being a unit test this can't really-actually check BWC so all it's really doing is feeding the output of `writeTo` of one into the `StreamInput` constructor of the other. If I'm wrong and there's some logic we should make sure is tested here please point me to it; this is an unorthodox thing I'm doing and I want to make sure all the bases are covered.

jakelandis commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/internal/TrialLicenseEra.java:
```diff
@@ -0,0 +1,93 @@
+ * that's already used a trial and let it expire. This class controls when we do that. The serialization of this class is designed to
+ * maintain compatibility with old-school Elasticsearch versions (specifically the {@link org.elasticsearch.Version} class).
+ */
+public class TrialLicenseEra implements ToXContentFragment, Writeable {
```
nit: (feel free ignore) I would prefer `TrialLicenseVersion` over `TrialLicenseEra` custom concept scoped versions will become (if not already) common place and I think `Era` will be the outlier w.r.t. naming.

gwbrown replied:
Hah, I used `Era` instead of `Version` to try to avoid arguments about introducing yet another Version class, but maybe those have already been had. I'm very open to calling it `Version`.

rjernst replied:
+1 to using the Version suffix, it's all the rage

jakelandis reviewed (approved):
LGTM

AthenaEryma committed:
Era -> Version

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into license/decouple-trial-from-version

rjernst commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/internal/TrialLicenseVersion.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    public TrialLicenseVersion(StreamInput in) throws IOException {
+        this.era = (in.readVInt() - 99) / 1000000; // Copied from the constructor of Version
```
In other new *Version forks from Version we have used the exact same serialization but diverging going forward (a bigger integer). Rather than try to reduce this down to the major version number from previous Version instances, can we pin the latest version id as the "current", and not do any manipulation to the id when read?

gwbrown replied:
Thanks, that is a better way of doing it - frankly this is the first thing that occurred to me and upon reflection that's obviously a better way of doing it. I'll make the updates.

rjernst commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/internal/TrialLicenseVersion.java:
```diff
@@ -0,0 +1,93 @@
+
+    // pkg-private for testing
+    String asVersionString() {
+        return this + ".0.0";
```
Can we use this bwc format only in bwc cases? Does the license maintain the version bit as a string or version id?

gwbrown replied:
A string, unfortunately:
https://github.com/elastic/elasticsearch/blob/878127ce7732e2ddab7c06fda375f3ef24b327f2/x-pack/plugin/core/src/main/java/org/elasticsearch/license/LicensesMetadata.java#L155
(⬆️ code is from `main`, to be clear).

Note this is `LicensesMetadata`, rather than in the license itself, so we only store one of these for the whole cluster. We can probably switch to emitting an integer, are there are any cases where we'd need an old version of ES to be able to parse XContent emitted by a newer version? The main case I'm aware of where we need to read a `LicensesMetadata` XContent is restoring a snapshot, and you can't restore a snapshot taken by a newer version than you're currently on AFAIK.

rjernst replied:
As long as we only upgrade the license metadata format after the min version of the cluster has stabilized in the upgrade, we should be ok.

rjernst reviewed (commented):
I have a couple thoughts on simplifying the serialized forms.

gwbrown commented on x-pack/plugin/core/src/test/java/org/elasticsearch/license/internal/TrialLicenseEraTests.java:
```diff
@@ -0,0 +1,98 @@
+
+    public void testVersionCanParseAllEras() {
+        for (int i = 2; i <= TrialLicenseEra.CURRENT.asInt(); i++) {
+            Version.fromString(new TrialLicenseEra(i).asVersionString());
```
I can remove this test, I found it necessary to gain enough confidence in this solution to be comfortable PR'ing it but it might not be valuable going forward. We might need to maintain the string format, if for some reason we need old versions of ES to be able to read the XContent emitted by newer ones (see the other thread on that point).

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into license/decouple-trial-from-version

AthenaEryma committed:
Address review feedback, specifically:

- Fix a few more leftover `era`s
- Remove serialization tests linked to `Version`
- Use integer as-is instead of extracting major version
- Disallow starting a trial license in a mixed-version cluster

rjernst commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/LicensesMetadata.java:
```diff
@@ -57,31 +57,31 @@ public class LicensesMetadata extends AbstractNamedDiffable<Metadata.Custom> imp
 
-    Version getMostRecentTrialVersion() {
-        return trialVersion;
+    TrialLicenseVersion getMostRecentTrialEra() {
```
leftover "era" terminology?

rjernst commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/ClusterStateLicenseService.java:
```diff
@@ -249,12 +250,12 @@ public ClusterState execute(ClusterState currentState) throws Exception {
                     Metadata currentMetadata = currentState.metadata();
                     LicensesMetadata licensesMetadata = currentMetadata.custom(LicensesMetadata.TYPE);
-                    Version trialVersion = null;
+                    TrialLicenseVersion trialEra = null;
```
leftover "era" terminology?

rjernst commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/StartBasicClusterTask.java:
```diff
@@ -78,8 +78,8 @@ public LicensesMetadata execute(
             }
-            Version trialVersion = currentLicensesMetadata != null ? currentLicensesMetadata.getMostRecentTrialVersion() : null;
-            updatedLicensesMetadata = new LicensesMetadata(selfGeneratedLicense, trialVersion);
+            TrialLicenseVersion trialEra = currentLicensesMetadata != null ? currentLicensesMetadata.getMostRecentTrialEra() : null;
```
leftover "era"

rjernst commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/StartupSelfGeneratedLicenseTask.java:
```diff
@@ -87,8 +87,8 @@ private ClusterState updateLicenseSignature(ClusterState currentState, LicensesM
         License selfGeneratedLicense = SelfGeneratedLicense.create(specBuilder, currentState.nodes());
-        Version trialVersion = currentLicenseMetadata.getMostRecentTrialVersion();
-        LicensesMetadata newLicenseMetadata = new LicensesMetadata(selfGeneratedLicense, trialVersion);
+        TrialLicenseVersion trialEra = currentLicenseMetadata.getMostRecentTrialEra();
```
leftover "era"

rjernst reviewed (approved):
LGTM. Only remaining nit is there are still quite a few uses of the "era" terminology in method and variable names.

AthenaEryma committed:
Yet more era->version

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into license/decouple-trial-from-version

AthenaEryma committed:
Remove another unnecessary test & another version->era

gwbrown commented:
Failure is https://github.com/elastic/elasticsearch/issues/97260

gwbrown commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into license/decouple-trial-from-version
